src/poll/responser.rs

```diff
@@ -0,0 +1,666 @@
+// TODO: this `#![allow(unused)]` lint is temporarily added. Will be removed sooner.
+#![allow(unused)]
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::{net::Ipv4Addr, slice::from_raw_parts_mut, sync::Arc, thread::spawn};
+
+use bitfield::bitfield;
+use libc::IFA_ADDRESS;
+use lockfree::queue::Queue;
+
+use crate::qp::QpContext;
+use crate::{device::QpType::RawPacket, Sge};
+use crate::{
+    device::{
+        ToCardWorkRbDescBuilder, ToCardWorkRbDescCommon, ToHostWorkRbDescAethCode,
+        ToHostWorkRbDescOpcode, ToHostWorkRbDescRead,
+    },
+    Device, Error,
+};
+
+/// A single MR that contains a free list of packet buffer. Each slot is 64 Byte.
+pub(crate) const ACKNOWLEDGE_BUFFER_SLOT: usize = 64;
+/// The size of the ack buffer.
+pub(crate) const ACKNOWLEDGE_BUFFER_SIZE: usize = ACKNOWLEDGE_BUFFER_SLOT * 1024;
+
+/// A interface that allows DescResponser to push the work descriptor to the device
+pub trait WorkDescriptorSender: Send + Sync {
+    fn push(&self, desc_builder: ToCardWorkRbDescBuilder) -> Result<(), Error>;
+}
+
+impl WorkDescriptorSender for Device {
+    fn push(&self, desc_builder: ToCardWorkRbDescBuilder) -> Result<(), Error> {
+        self.send_work_desc(desc_builder)
+    }
+}
+
+/// Command about ACK and NACK
+/// Typically, the message is sent by checker thread, which is responsible for checking the packet ordering.
+pub(crate) struct RespAckCommand {
+    pub(crate) dpqn: u32,
+    pub(crate) msn: u32,
+    pub(crate) last_retry_psn: Option<u32>,
+}
+
+/// Command about read response
+pub(crate) struct RespReadRespCommand {
+    pub(crate) desc: ToHostWorkRbDescRead,
+}
+
+/// The response command sent by other threads
+///
+/// Currently, it supports two types of response:
+/// * Acknowledge(ack,nack)
+/// * Read Response
+pub(crate) enum RespCommand {
+    Acknowledge(RespAckCommand), // Acknowledge or Negative Acknowledge
+    ReadResponse(RespReadRespCommand),
+}
+
+/// A thread that is responsible for sending the response to the other side
+pub(crate) struct DescResponser {
+    thread: std::thread::JoinHandle<()>,
+}
+
+impl DescResponser {
+    pub fn new(
+        device: Arc<dyn WorkDescriptorSender>,
+        recving_queue: std::sync::mpsc::Receiver<RespCommand>,
+        ack_buffers: AcknowledgeBuffer,
+        qp_table: Arc<RwLock<HashMap<u32, QpContext>>>,
+    ) -> Self {
+        let thread = spawn(|| Self::working_thread(device, recving_queue, ack_buffers, qp_table));
+        Self { thread }
+    }
+
+    fn working_thread(
+        device: Arc<dyn WorkDescriptorSender>,
+        recving_queue: std::sync::mpsc::Receiver<RespCommand>,
+        ack_buffers: AcknowledgeBuffer,
+        qp_table: Arc<RwLock<HashMap<u32, QpContext>>>,
```

Comment:
use newtype to define u32 as a new type to represent QPN will be better. QPN is 24bits, and infact our hardware can't support that much, maybe 128 is enough, so when creating a QPN, we should check it if it exceed max limit.

Reply:
Yes. I have replaced it with a new type. See the next commit.